File: android/src/vc/duvido/FacebookBridge.cpp
#include <vc.duvido.FacebookBridge.hpp>
#include <com.facebook.AccessToken.hpp>
#include <java.lang.String.hpp>
#include <core/duvido.hpp>
#include <core/facebook.hpp>

using namespace vc::duvido;
using namespace java::lang;

void FacebookBridge::onLoginSuccess() {
    // TODO
}

void FacebookBridge::onLoginError() {
    // TODO
}

void FacebookBridge::onLoginCancel() {
    // TODO
}

void FacebookBridge::onAccessTokenReceived(const String& token) {
    if (token.isNull()) return;
    // This method must be invoked on duvido's thread. QueuedConnection will launch a QEvent on the right thread.
    Duvido::staticMetaObject.invokeMethod(::duvido, "setToken", Qt::QueuedConnection, Q_ARG(QString, token));
}

File: android/src/vc/duvido/FacebookBridge.java
package vc.duvido;

import android.app.Activity;
import android.content.Intent;

import com.facebook.AccessToken;
import com.facebook.AccessTokenTracker;
import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.FacebookSdk;
import com.facebook.login.LoginManager;
import com.facebook.login.LoginResult;

import java.util.ArrayList;

public class FacebookBridge {

    public static final int RESULT_WEBLOGIN = 10;

    private static CallbackManager callbackManager;

    public static void initialize() {
        FacebookSdk.sdkInitialize(DuvidoActivity.getInstance());
        FacebookSdk.setIsDebugEnabled(true);

        new AccessTokenTracker() {
            @Override
            protected void onCurrentAccessTokenChanged(AccessToken oldAccessToken, AccessToken newAccessToken) {
                if (newAccessToken != null)
                    onAccessTokenReceived(newAccessToken.getToken());
            }
        };

        callbackManager = CallbackManager.Factory.create();

        LoginManager.getInstance().registerCallback(callbackManager, new FacebookCallback<LoginResult>() {
           @Override
           public void onSuccess(LoginResult loginResult) {
               onLoginSuccess();
           }

           @Override
           public void onCancel() {
               onLoginCancel();
           }

           @Override
           public void onError(FacebookException e) {
               onLoginError();
           }
       });
    }

    public static void submitActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
        case RESULT_WEBLOGIN:
            if (resultCode == Activity.RESULT_OK) {
                onAccessTokenReceived(data.getStringExtra("token"));
            }
            break;
        default:
            callbackManager.onActivityResult(requestCode, resultCode, data);
        }
    }

    public static void login() {

        if (DuvidoActivity.getInstance().hasFacebookApp()) {
            ArrayList<String> perms = new ArrayList<String>();
            perms.add("user_friends");
            perms.add("public_profile");
            perms.add("user_birthday");
            perms.add("email");
            LoginManager.getInstance().logInWithReadPermissions(DuvidoActivity.getInstance(), perms);
        }
        else {
            DuvidoActivity.getInstance().startWebLogin();
        }
    }

    public native static void onLoginSuccess();
    public native static void onLoginCancel();
    public native static void onLoginError();
    public native static void onAccessTokenReceived(String token);

}
